dsl/src/commonMain/kotlin/com/freeletics/flowredux/dsl/internal/CollectStateInStateSideEffectBuilder.kt

```diff
@@ -0,0 +1,60 @@
+package com.freeletics.flowredux.dsl.internal
+
+import com.freeletics.flowredux.SideEffect
+import com.freeletics.flowredux.GetState
+import com.freeletics.flowredux.dsl.FlatMapPolicy
+import com.freeletics.flowredux.dsl.InStateObserverHandler
+import com.freeletics.flowredux.dsl.flow.flatMapWithPolicy
+import com.freeletics.flowredux.dsl.flow.whileInState
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.map
+
+/**
+ * A builder to create a [SideEffect] that observes a Flow<T> as long as the redux store is in
+ * the given state. We use is instance of check to check if a new state has been reached and Flow<T>
+ * is closed.
+ */
+@FlowPreview
+@ExperimentalCoroutinesApi
+internal class CollectStateInStateBuilder<T, InputState : S, S : Any, A : Any>(
+    private val isInState: (S) -> Boolean,
+    private val flowBuilder: (Flow<InputState>) -> Flow<T>,
+    private val flatMapPolicy: FlatMapPolicy,
+    private val handler: InStateObserverHandler<T, InputState, S>
+) : InStateSideEffectBuilder<InputState, S, A>() {
+
+    @Suppress("unchecked_cast")
+    override fun generateSideEffect(): SideEffect<S, Action<S, A>> {
+        return { actions: Flow<Action<S, A>>, getState: GetState<S> ->
+            actions.whileInState(isInState, getState) { inStateActions ->
+                inStateActions.map { getState() as InputState }
```

Comment:
> So while it works, I was wondering if instead of subscribing to the Actions directly collect the StateFlow from FlowReduxStateMachine?

We can't do that ourselves because we don't have the StateFlow available here. A user could do it with the existing `collectWhileInState`, but I think doing that is weird and you need to cast yourself.

Extracting parts of this into a function to make readability better makes sense

Reply:
9abaac9 this should improve the readability and also adds some comments to the extracted function.